### PR TITLE
Port to Hephaistos v2

### PR DESCRIPTION
**Pull request by jglrxavpok**
Hephaistos v2 introduces the following changes:
* Support for 1.17+ chunks. 0..255 is no longer the required range for Y values.
* Immutability of NBT types
* RegionFile can be used in-memory
* Hosted on Maven Central
* Various bug fixes (SNBT string parsing is one)
* Support for 1.18 worlds

A migration guide is available for Minestom users at the following address: [https://github.com/jglrxavpok/Hephaistos/blob/master/docs/migration_guide_v1_v2.md](https://github.com/jglrxavpok/Hephaistos/blob/master/docs/migration_guide_v1_v2.md)

### CODE REVIEW

File: src/main/java/net/minestom/server/entity/Entity.java [hunk 1]
```diff
@@ -1565,7 +1566,7 @@ private Duration getSynchronizationCooldown() {
 
     @Override
     public <T> @Nullable T getTag(@NotNull Tag<T> tag) {
-        return tag.read(nbtCompound);
+        return tag.read(nbtCompound.toCompound());
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Entity.java:1569]: I assume that the conversion is unnecessary?

File: src/main/java/net/minestom/server/entity/Metadata.java [hunk 2]
```diff
@@ -246,6 +246,36 @@ public void write(@NotNull BinaryWriter writer) {
             //noinspection unchecked
             return (Value<T>) value.withValue(reader);
         }
+
+        @NotNull
+        public Value<T> getMetaValue() {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Metadata.java:251]: Is it used anywhere? Or a rebase artifact?

File: src/main/java/net/minestom/server/entity/Metadata.java [hunk 3]
```diff
@@ -246,6 +246,36 @@ public void write(@NotNull BinaryWriter writer) {
...
+            return value;
+        }
+    }
+
+    private static <T> Value<T> getCorrespondingNewEmptyValue(int type) {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Metadata.java:256]: Same as above

File: src/main/java/net/minestom/server/item/metadata/CrossbowMeta.java [hunk 4]
```diff
@@ -94,11 +96,11 @@ public Builder projectile(@NotNull ItemStack projectile) {
             this.projectile1 = projectile;
             this.triple = false;
 
-            NBTList<NBTCompound> chargedProjectiles = new NBTList<>(NBTTypes.TAG_Compound);
+            List<NBTCompound> chargedProjectiles = new LinkedList<>();
```
Comment (TheMode) [inline src/main/java/net/minestom/server/item/metadata/CrossbowMeta.java:99]: Should be an `ArrayList`

File: src/main/java/net/minestom/server/item/ItemMetaBuilder.java [hunk 5]
```diff
@@ -85,10 +87,11 @@
...
         this.lore = new ArrayList<>(lore);
         handleCompound("display", nbtCompound -> {
-            final NBTList<NBTString> loreNBT = new NBTList<>(NBTTypes.TAG_String);
-            for (Component line : lore) {
-                loreNBT.add(new NBTString(GsonComponentSerializer.gson().serialize(line)));
-            }
+            final NBTList<NBTString> loreNBT = NBT.List(NBTType.TAG_String,
+                    lore.stream()
+                            .map(line -> new NBTString(GsonComponentSerializer.gson().serialize(line)))
+                            .collect(Collectors.toList())
```
Comment (TheMode) [inline src/main/java/net/minestom/server/item/ItemMetaBuilder.java:93]: All those could be replaced by the newer `#toList()` allocating an immutable list

**Review by TheMode — CHANGES_REQUESTED** [review]
A few nitpicks

File: src/main/java/net/minestom/server/entity/Entity.java (hunk 1 above)
Comment (jglrxavpok) [reply src/main/java/net/minestom/server/entity/Entity.java:1569]: Indeed, will fix in next commit

File: src/main/java/net/minestom/server/entity/Metadata.java (hunk 2 above)
Comment (jglrxavpok) [reply src/main/java/net/minestom/server/entity/Metadata.java:251]: Artifact, will fix in next commit

File: src/main/java/net/minestom/server/item/ItemMetaBuilder.java (hunk 5 above)
Comment (jglrxavpok) [reply src/main/java/net/minestom/server/item/ItemMetaBuilder.java:93]: Will fix in next commit

File: src/main/java/net/minestom/server/item/metadata/CrossbowMeta.java (hunk 4 above)
Comment (jglrxavpok) [reply src/main/java/net/minestom/server/item/metadata/CrossbowMeta.java:99]: Will fix in next commit

**Review by TheMode — APPROVED** [review]

**Pull request merged**